packages/entity/src/utils/EntityPrivacyUtils.ts

```diff
@@ -0,0 +1,311 @@
+ * @param queryContext - query context in which to perform the check
+ */
+export async function canViewerDeleteAsync<
+  TFields extends object,
+  TID extends NonNullable<TFields[TSelectedFields]>,
+  TMViewerContext extends ViewerContext,
+  TEntity extends Entity<TFields, TID, TMViewerContext, TSelectedFields>,
+  TPrivacyPolicy extends EntityPrivacyPolicy<
+    TFields,
+    TID,
+    TMViewerContext,
+    TEntity,
+    TSelectedFields
+  >,
+  TSelectedFields extends keyof TFields = keyof TFields
+>(
+  entityClass: IEntityClass<
+    TFields,
+    TID,
+    TMViewerContext,
+    TEntity,
+    TPrivacyPolicy,
+    TSelectedFields
+  >,
+  sourceEntity: TEntity,
+  queryContext: EntityQueryContext = sourceEntity
+    .getViewerContext()
+    .getViewerScopedEntityCompanionForClass(entityClass)
+    .getQueryContextProvider()
+    .getQueryContext()
+): Promise<boolean> {
+  return await canViewerDeleteInternalAsync(
+    entityClass,
+    sourceEntity,
+    /* cascadingDeleteCause */ null,
+    queryContext
+  );
+}
+
+async function canViewerDeleteInternalAsync<
+  TFields extends object,
+  TID extends NonNullable<TFields[TSelectedFields]>,
+  TMViewerContext extends ViewerContext,
+  TEntity extends Entity<TFields, TID, TMViewerContext, TSelectedFields>,
+  TPrivacyPolicy extends EntityPrivacyPolicy<
+    TFields,
+    TID,
+    TMViewerContext,
+    TEntity,
+    TSelectedFields
+  >,
+  TSelectedFields extends keyof TFields = keyof TFields
+>(
+  entityClass: IEntityClass<
+    TFields,
+    TID,
+    TMViewerContext,
+    TEntity,
+    TPrivacyPolicy,
+    TSelectedFields
+  >,
+  sourceEntity: TEntity,
+  cascadingDeleteCause: EntityCascadingDeletionInfo | null,
+  queryContext: EntityQueryContext
+): Promise<boolean> {
+  const viewerContext = sourceEntity.getViewerContext();
+  const entityCompanionProvider = viewerContext.entityCompanionProvider;
+
+  const viewerScopedCompanion = sourceEntity
+    .getViewerContext()
+    .getViewerScopedEntityCompanionForClass(entityClass);
+  {
+    const privacyPolicy = viewerScopedCompanion.entityCompanion.privacyPolicy;
+    const evaluationResult = await asyncResult(
+      privacyPolicy.authorizeDeleteAsync(
+        sourceEntity.getViewerContext(),
+        queryContext,
+        { cascadingDeleteCause },
+        sourceEntity,
+        viewerScopedCompanion.getMetricsAdapter()
+      )
+    );
+    if (!evaluationResult.ok) {
+      if (evaluationResult.reason instanceof EntityNotAuthorizedError) {
+        return false;
+      } else {
+        throw evaluationResult.reason;
+      }
+    }
+  }
+
+  const newCascadingDeleteCause = {
+    entity: sourceEntity,
+    cascadingDeleteCause,
+  };
+
+  // Take entity X which is proposed to be deleted, look at inbound edges (entities that reference X).
+  // These inbound edges are the entities that will either get deleted
+  // or updated with null based on the edge definiton when entity X is deleted.
```

Comment:
```suggestion
  // or updated with null foreign keys based on the edge definition when entity X is deleted.
```

Reply:
Updated to roughly what is suggested, though foreign keys are a relational concept while entity is theoretically db-agnostic so I used a different phrasing. Let me know if you think it still needs alteration.

Reply:
What I wanted to clarify is that the FK fields of the referencing entities will be set to null. Basically, for "Updated with null" to be more precise and say what exactly will be set to null.

Reply:
Updated again. Thanks for the clarification.